### DISCUSSION TITLE
Crossfade/normalize

### DISCUSSION THREAD

**Discussion by SpeedyCharly** [savonet/liquidsoap#2128] [Q&A]
### Discussed in https://github.com/savonet/liquidsoap/discussions/2106

<div type='discussions-op-text'>

<sup>Originally posted by **SpeedyCharly** December 14, 2021</sup>
I've posted the same question on Azuracast github, but no one seems able to respond;

I've taken to studying the Liquidsoap logs to familiarize myself with Azura's operations. Here is a piece that comes up often in my case, and I don't quite understand why...
```
2021/11/25 22:24:25 [next_song:3] Prepared "/var/azuracast/stations/radio_des_festivals/media/Country Franco/La Cabane À Sucre - Viateur Caron (New).mp3" (RID 5).
2021/11/25 22:24:25 [cue_next_song:3] Cueing in...
2021/11/25 22:24:25 [crossfade_0:3] Analysis: -80.616358dB / -24.460022dB (1.44s / 1.44s)
2021/11/25 22:24:25 [crossfade_0:3] new >= old + margin, old <= medium and new <= high.
2021/11/25 22:24:25 [crossfade_0:3] Do not fade if it's already very low.
2021/11/25 22:24:25 [crossfade_0:3] Transition: crossed, no fade.
2021/11/25 22:24:25 [lang:3] AzuraCast Raw Response: annotate:title="Le temps qu'il nous reste",artist="Claude Martel",duration="252.00",song_id="b895d5adb8468a0d7b55b7d91646992b",media_id="8587",liq_amplify="0.00dB",playlist_id="2":/var/azuracast/stations/radio_des_festivals/media/Country Franco/10 Le temps quil nous reste.mp3
````
All of my playlists are normalized to the same target volume with MP3 Gain and, Normalize is also set in my AzuraCast config. Therefore, such occurrences shouldn't exist. From the volume analysis above, it seems that Liquidsoap doesn't recognize the MP3 Gain metadata? If so, would another normalizer be a better choice?
![145714206-5c0c8558-efc4-430a-b97e-b3539448ea78](https://user-images.githubusercontent.com/66107794/146020756-b0da3c0b-1908-41f4-82fd-f2b85f77f73e.jpg)

As an aside, one more question; does Liqudsoap natively truncate leading/ending silence on media as part of the crossfade process? I don't see a switch [as in SamBroadcaster] to turn this on or off.
</div>

**Comment by toots** [comment]
The analysis for crossfade is done on the end and beginning of tracks. It is not representative of the loudness of the whole track but only of the portion that needs to be used to compute the crossfade transition. So, a loud song that already ends with a long fade out will in fact show up with low DB level in the analysis.

Handling of replaygain and normalization is not enabled by default and is documented here: https://www.liquidsoap.info/doc-2.0.1/replay_gain.html

More generally, I can recommend the liquidsoap book at as starting point: https://www.liquidsoap.info/doc-2.0.1/book.html it does answer a lot of basic questions about how liquidsoap works.

**Comment by SpeedyCharly** [comment]
Replay Gain has been activated of late in Azuracast, but it fails every time on unspecified error 

```
2022/01/05 06:38:41 [extract.replaygain:4] ReplayGain extraction by /var/azuracast/.opam/4.12.0/lib/liquidsoap/share/liquidsoap/2.0.2/bin/extract-replaygain failed with: 
2022/01/05 06:38:41 [next_song:4] Queued 1 requests
```

Long fades and leading/ending silences that may adversely affect proper crossfading are handled in other apps by truncating silences and  setting a certain dB level to trigger the crossfade process.  Such doesn't appear to be the case in LS as we often hear songs fading to and up from dead silence....  

Respectfully, methinks there is much room for improvement in that department.

**Reply by smimram** [reply]
What happens if you manually try to run
```
/var/azuracast/.opam/4.12.0/lib/liquidsoap/share/liquidsoap/2.0.2/bin/extract-replaygain some-file.mp3
```
(where you should replace `some-file.mp3` with a valid file)?

**Reply by SpeedyCharly** [reply]
COMMAND NOT FOUND

**Reply by smimram** [reply]
Does the file `/var/azuracast/.opam/4.12.0/lib/liquidsoap/share/liquidsoap/2.0.2/bin/extract-replaygain` exist in the first place? Otherwise, it looks like a problem with your installation...

**Reply by SpeedyCharly** [reply]
I've inherited this install from another operator, whom I no longer have contact with, and who had done a non-standard installation at the outset.  As I see it, /var/azuracast/ is an empty folder and it took me a while to even locate the docker files to do simple updates...

**Reply by SpeedyCharly** [reply]
Silence......  Does that mean that I am unworthy of further assistance du to my non-standard installation?

**Reply by toots** [reply]
Well if the installation is broken, there isn't much that we can do.

**Comment by SpeedyCharly** [comment]
nothing broken... I've been running this install for 2 years ... it's just that the install wasn't done the standard way and resources are displaced to different folders. Not a Linux tech here so I can't fix it on my own..

Things can't be that bad cuz Rolling releases get applied fine...
